From 0b25a7b4d6969747e558ebeb349fc957546d9b86 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Wed, 5 Mar 2025 16:08:16 +0100
Subject: [PATCH 1635/1695] clk: stm32mp21: dont modify reserved RCC RIF
 registers

Add safeguard to prevent writing to reserved RIFF RCC registers.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c9d5bd1ad17f3a4ca32411e1e488fa67aacc1473 clk: stm32mp21: dont modify reserved RCC RIF registers]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: Ie7abbc59f1791592ec9aa4bc9b11a9af74a3449a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/441933
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 32 ++++++++++++++++++++++++++++----
 1 file changed, 28 insertions(+), 4 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index bdd86991a..2fcdfa4f4 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -4481,6 +4481,26 @@ static struct clk_stm32_priv stm32mp21_clock_data = {
 	.is_critical		= clk_stm32_clock_is_critical,
 };
 
+static bool is_rcc_rif_reserved(unsigned int id)
+{
+	switch (id) {
+	case 72:
+	case 79:
+	case 80:
+	case 81:
+	case 82:
+	case 89:
+	case 99:
+	case 100:
+	case 105:
+	case 107:
+	case 111:
+		return true;
+	default:
+		return false;
+	}
+}
+
 static TEE_Result handle_available_semaphores(void)
 {
 	struct stm32_clk_platdata *pdata = &stm32mp21_clock_pdata;
@@ -4494,7 +4514,8 @@ static TEE_Result handle_available_semaphores(void)
 
 		index = i / 32;
 
-		if (!(BIT(i % 32) & pdata->conf_data.access_mask[index]))
+		if (is_rcc_rif_reserved(i) ||
+		    (!(BIT(i % 32) & pdata->conf_data.access_mask[index])))
 			continue;
 
 		cidcfgr = io_read32(pdata->rcc_base + RCC_CIDCFGR(i));
@@ -4535,8 +4556,10 @@ static TEE_Result apply_rcc_rif_config(bool is_tdcid)
 	if (is_tdcid) {
 		for (i = 0; i < RCC_NB_RIF_RES; i++) {
 			index = i / 32;
-			if (!(BIT(i % 32) &
-			      pdata->conf_data.access_mask[index]))
+
+			if (is_rcc_rif_reserved(i) ||
+			    (!(BIT(i % 32) &
+			       pdata->conf_data.access_mask[index])))
 				continue;
 
 			/*
@@ -4570,7 +4593,8 @@ static TEE_Result apply_rcc_rif_config(bool is_tdcid)
 	for (i = 0; i < RCC_NB_RIF_RES; i++) {
 		index = i / 32;
 
-		if (!(BIT(i % 32) & pdata->conf_data.access_mask[index]))
+		if (is_rcc_rif_reserved(i) ||
+		    !(BIT(i % 32) & pdata->conf_data.access_mask[index]))
 			continue;
 
 		io_clrsetbits32(pdata->rcc_base + RCC_CIDCFGR(i),
-- 
2.47.2

